src/main/java/com/uber/jenkins/phabricator/coverage/XmlCoverageProvider.java

```diff
@@ -0,0 +1,408 @@
+            localDb = dbf.newDocumentBuilder();
+        } catch (ParserConfigurationException e) {
+            e.printStackTrace();
+        }
+        db = localDb;
+        cc = new CoverageCounters();
+    }
+
+    @Override
+    protected void computeCoverage() {
+        try {
+            parse(includeFiles, coverageReports);
+        } catch (SAXException | IOException e) {
+            e.printStackTrace();
+        }
+        // Aggregate coverage metrics
+        metrics = new CodeCoverageMetrics(
+                cc.pkg.getPrecent(),
+                cc.file.getPrecent(),
+                cc.cls.getPrecent(),
+                cc.method.getPrecent(),
+                cc.line.getPrecent(),
+                cc.branch.getPrecent(),
+                cc.line.covered,
+                cc.line.covered + cc.line.missed
+        );
+    }
+
+    private void parse(Set<String> includeFiles, Set<File> reports) throws SAXException, IOException {
+        if (db != null) {
+            for (File file : reports) {
+                try (InputStream is = new FileInputStream(file)) {
+                    Document document = db.parse(is);
+                    for (XmlCoverageHandler xmlCoverageHandler : xmlCoverageHandlers) {
+                        if (xmlCoverageHandler.isApplicable(document)) {
+                            xmlCoverageHandler.parseCoverage(document, includeFiles, cc, lineCoverage);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private abstract static class XmlCoverageHandler {
+
+        abstract boolean isApplicable(Document document);
+
+        abstract void parseCoverage(
+                Document document,
+                Set<String> includeFiles,
+                CoverageCounters cc,
+                Map<String, List<Integer>> lineCoverage);
+
+        void computeLineCoverage(
+                Map<String, SortedMap<Integer, Integer>> internalCounts,
+                Map<String, List<Integer>> lineCoverage) {
+            for (Map.Entry<String, SortedMap<Integer, Integer>> entry : internalCounts.entrySet()) {
+                List<Integer> sortedCounts = new ArrayList<>();
+                int startIndex = 1;
+                for (Map.Entry<Integer, Integer> line : entry.getValue().entrySet()) {
+                    for (int i = startIndex; i < line.getKey(); i++) {
+                        sortedCounts.add(null);
+                        startIndex++;
+                    }
+                    sortedCounts.add(line.getValue());
+                    startIndex++;
+                }
+                lineCoverage.put(entry.getKey(), sortedCounts);
+            }
+        }
+
+        static int getIntValue(Node node, String attributeName) {
+            return Integer.parseInt(node.getAttributes().getNamedItem(attributeName).getTextContent());
+        }
+    }
+
+    private static class CoberturaXmlCoverageHandler extends XmlCoverageHandler {
```

Comment:
I assume below handlers are brought over from corresponding jenkins plugin, should we add the licenses for those?

Reply:
This was written as a clean room implementation and does not follow the upstream version at all